aten/src/ATen/detail/CUDAHooksInterface.h

```diff
@@ -0,0 +1,114 @@
+#pragma once
+
+#include <ATen/Registry.h>
+#include <ATen/Generator.h>
+#include <ATen/Error.h>
+#include <ATen/Allocator.h>
+
+// Forward declare these CUDA types here to avoid including CUDA headers in
+// ATen headers, which would make ATen always require CUDA to build.
+struct THCState;
+struct CUstream_st;
+typedef struct CUstream_st *cudaStream_t;
+struct cudaDeviceProp;
+
+namespace at {
+  class Context;
+}
+
+// NB: Class must live in at due to limitations of Registry.h
+namespace at {
+
+// The CUDAHooksInterface is an omnibus interface for any CUDA functionality
+// which we may want to call into from CPU code (and thus must be dynamically
+// dispatched, to allow for separate compilation of CUDA code).  How do I
+// decide if a function should live in this class?  There are two tests:
+//
+//  1. Does the *implementation* of this function require linking against
+//     CUDA libraries?
+//
+//  2. Is this function *called* from non-CUDA ATen code?
```

Comment:
I think those two points are not enough to justify having all the functions you put there. Why would CPU-only code care how many CUDA devices there are? Why would it ever want to get a current CUDA stream?

Also, having min eps for batch norm in here is very weird too. We should just have a single entry point for CUDA dispatch in the CPU library, and handle choosing a CUDA impl internally in `ATen_cuda`. Ofc it doesn't have to happen in this PR, and I guess C10 might resolve the issue anyway.

Reply:
Right, so I think eventually we want way less functions in this interface. When I put things in here, I didn't bother "changing" any of ATen's API; if there was a "CPU" part of ATen which needed some CUDA information, I tossed it in. I think that many of the examples you provide can be eliminated.  I absolutely agree with your point about batchnorm min-epsilon.

I do want to push on the matter of "should ATen have a getNumGPUs function" a little more. I think there's a pretty strong case to be made for putting making an "ATen CUDA interface", which you can only link against if you actually link against `libATen_cuda.so`. However, I don't think it is a 100% slam dunk. Let's think about an average Python script which supports both CPU and CUDA execution. The same Python script works no matter if you run it with PyTorch compiled with CPU+CUDA, or PyTorch with CPU only. The reason, of course, is that you don't actually exercise the CUDA functionality unless you *actually call it*, so as long as you don't pass `--cuda` to the trainer script it will happily use the CPU functionality. Now let's consider the corresponding situation in C++ land. If you link against a `libATen_cuda.so` only symbol, you MUST have an install of PyTorch which has `libATen_cuda.so` available (in all its 300M glory) (which in turn must have `libcudart.so`), UNLESS we distribute some sort of `stub/libATen_cuda.so` which defines all of the public symbols of libATen_cuda.so but with trivial implementations of them. On the other hand, if the symbols were defined in `libATen.so` and dynamically dispatched to `libATen_cuda.so`, everything would work perfectly fine. Now, to be clear, I am not sure what the correct decision here is, but I don't think it is trivial.

Reply:
(Blah, actually, batchnorm min-epsilon is not so easy to expunge: the problem is that the current dispatch logic doesn't know how dispatch first to CUDA, and *then* to the autograd-enabled sub-implementation. If I give anything `dispatch: CUDA` then I immediately lose pass through and have to implement the backwards manually. Don't want to do that.)